File: python/simplesauce/session.py
import os
from selenium import webdriver
from selenium.webdriver.remote.remote_connection import RemoteConnection
from simplesauce.options import SauceOptions


SAUCE_USERNAME = os.getenv('SAUCE_USERNAME', None)
SAUCE_ACCCESS_KEY = os.getenv('SAUCE_ACCESS_KEY', None)

us_ondemand = 'ondemand.saucelabs.com'
eu_ondemand = 'ondemand.eu-central-1.saucelabs.com'

US_SAUCE_DC_URL = 'https://{}:{}@{}/wd/hub'.format(SAUCE_USERNAME, SAUCE_ACCCESS_KEY, us_ondemand)
EU_SAUCE_DC_URL = 'https://{}:{}@o{}/wd/hub'.format(SAUCE_USERNAME, SAUCE_ACCCESS_KEY, eu_ondemand)


class SauceSession():

    def __init__(self, **kwargs):

        # TODO: flesh this out
        self.options = SauceOptions() if not kwargs else kwargs
        self.remote_url = "https://ondemand.saucelabs.com/wd/hub"
        self.driver = {}

    def start(self):
        caps = self.options

        executor = RemoteConnection(self.remote_url, resolve_ip=False)
        self.driver = webdriver.Remote(
            command_executor=executor,
            desired_capabilities=caps,
            keep_alive=True
        )
        return self.driver

    def stop(self):
        self.driver.quit()

File: ruby/lib/sauce_bindings/session.rb
# frozen_string_literal: true

require 'sauce_whisk'
require 'selenium-webdriver'

module SauceBindings
  class Session
    DATA_CENTERS = {US_WEST: 'ondemand.us-west-1.saucelabs.com',
                    US_EAST: 'ondemand.us-east-1.saucelabs.com',
                    EU_VDC: 'ondemand.eu-central-1.saucelabs.com'}.freeze

    attr_writer :url
    attr_reader :driver, :options, :data_center
    attr_accessor :http_client, :listener

    def initialize(options = nil, data_center: nil, http_client: nil, listener: nil)
      @options = options || Options.new
      @http_client = http_client
      @listener = listener

      @username = ENV['SAUCE_USERNAME']
      @access_key = ENV['SAUCE_ACCESS_KEY']
      self.data_center = data_center || :US_WEST
    end

    def start
      raise ArgumentError, "needs username; use `ENV['SAUCE_USERNAME']`" unless @username
      raise ArgumentError, "needs access_key; use `ENV['SAUCE_ACCESS_KEY']`" unless @access_key

      @driver = Selenium::WebDriver.for :remote, to_selenium
    end

    def stop(result)
      return if @driver.nil?

      SauceWhisk::Jobs.change_status(@driver.session_id, result)
      @driver.quit
    end

    def data_center=(data_center)
      unless DATA_CENTERS.key?(data_center)
        msg = "#{data_center} is an invalid data center; specify :US_WEST, :US_EAST or :EU_VDC"
        raise ArgumentError, msg
      end

      SauceWhisk.data_center = data_center
      @data_center = data_center
    end

    def url
      @url ||= "https://#{@username}:#{@access_key}@#{DATA_CENTERS[data_center]}:443/wd/hub"
    end

    def to_selenium
      caps = {url: url, desired_capabilities: options.capabilities}
      caps[:listener] = listener if listener
      caps[:http_client] = http_client if http_client
      caps
    end
  end
end